yt/frontends/nc4_cm1/io.py

```diff
@@ -0,0 +1,62 @@
+import numpy as np
+
+from yt.utilities.file_handler import NetCDF4FileHandler
+from yt.utilities.io_handler import BaseIOHandler
+
+
+class CM1IOHandler(BaseIOHandler):
+    _particle_reader = False
+    _dataset_type = "cm1"
+
+    def __init__(self, ds):
+        self.filename = ds.filename
+        self._handle = NetCDF4FileHandler(self.filename)
+        super(CM1IOHandler, self).__init__(ds)
+
+    def _read_particle_coords(self, chunks, ptf):
+        # This needs to *yield* a series of tuples of (ptype, (x, y, z)).
+        # chunks is a list of chunks, and ptf is a dict where the keys are
+        # ptypes and the values are lists of fields.
+        pass
+
+    def _read_particle_fields(self, chunks, ptf, selector):
+        # This gets called after the arrays have been allocated.  It needs to
+        # yield ((ptype, field), data) where data is the masked results of
+        # reading ptype, field and applying the selector to the data read in.
+        # Selector objects have a .select_points(x,y,z) that returns a mask, so
+        # you need to do your masking here.
+        pass
+
+    def _read_fluid_selection(self, chunks, selector, fields, size):
```

Comment:
Important note. As a frontend user, I did not understand what this method was for before it was too late. We now have a bug in the AMRVAC frontend (unfortunately not solved yet) here https://github.com/yt-project/yt/issues/2720

I suggest we test this method with a similar script as found in the bug report.

Reply:
I just tested this out and it does look like it is an issue here too -- when making a slice plot with a cut region as the data source I get similar behavior as in #2720

Reply:
I'll look into it some more and see what I find... but I don't think this has to block this PR. I can open an issue if this PR gets merged before I figure it out?

Reply:
Yeah it would be unfair to hold this PR just fo this. A new issue wouldn't be necessary if the bug is similar enough to the one in AMRVAC, but you get the opportunity to add a warning or a NotImplementedError somewhere before this is merged :)

Reply:
@matthewturk and @n-claes, I think this thread might interest you